src/lib/Interpreter.hs

```diff
@@ -277,24 +274,34 @@ matchUPats _ _ = error "mismatched lengths"
 
 -- XXX: It is a bit shady to unsafePerformIO here since this runs during typechecking.
 -- We could have a partial version of the interpreter that fails when any IO is to happen.
-indices :: EnvReader m => Type n -> m n [Atom n]
-indices ty = fmap fromListE $ flip runScopedT1 (mempty :: IxCache n) $ do
-    env <- unsafeGetEnv
-    Distinct <- getDistinct
-    ix <- simplifiedIxInstance ty
-    let IdxRepVal size = unsafePerformIO $ runInterpM env $ evalMethod ix simpleIxSize UnitVal
-    return $ ListE $ unsafePerformIO $ runInterpM env $
-      forM [0..size - 1] $ evalMethod ix simpleUnsafeFromOrdinal . IdxRepVal
-    where
-      evalMethod ix method x = case method ix of
-        Abs decls lam -> do
-          evalDecls decls $ evalExpr $ App (Lam lam) $ sinkFromTop x NE.:| []
+liftBuilderInterp
+  :: EnvReader m
+  => (forall l. (Emits l, DExt n l) => BuilderM l (Atom l))
+  -> m n (Atom n)
+liftBuilderInterp cont = do
+  block <- buildBlockSimplified cont
+  env <- unsafeGetEnv
+  Distinct <- getDistinct
+  return $ unsafePerformIO $ runInterpM env $ evalBlock block
+{-# SCC liftBuilderInterp #-}
 
-      runInterpM :: Distinct n => Env n -> InterpM n n a -> IO a
-      runInterpM bindings cont =
-        runEnvReaderT bindings $ runSubstReaderT idSubst $ runInterpM' cont
+runInterpM :: Distinct n => Env n -> InterpM n n a -> IO a
+runInterpM bindings cont =
+  runEnvReaderT bindings $ runSubstReaderT idSubst $ runInterpM' cont
+{-# SCC runInterpM #-}
+
+indices :: EnvReader m => IxType n -> m n [Atom n]
+indices ixTy = do
+  ~(IdxRepVal size) <- interpIndexSetSize ixTy
+  forM [0..size-1] \i -> interpIntToIndex ixTy $ IdxRepVal i
 {-# SCC indices #-}
 
+interpIndexSetSize :: EnvReader m => IxType n -> m n (Atom n)
+interpIndexSetSize ixTy = liftBuilderInterp $ indexSetSize $ sink ixTy
+
+interpIntToIndex :: EnvReader m => IxType n -> Atom n -> m n (Atom n)
+interpIntToIndex ixTy i = liftBuilderInterp $ intToIndex (sink ixTy) (sink i)
```

Comment:
Good idea

Reply:
Actually this turns out to be a bit tricky because `liftBuilderInterp` requires an `Atom` result since it does simplification too. We can still do it if we want, but it doesn't seem urgent.